File: lib/awtrix.ts
import type {
  AwtrixCustomAppPayload,
  AwtrixNotifyPayload,
  DeviceConfig,
  PushResult,
  Question,
} from "@/types";

const REQUEST_TIMEOUT_MS = 4000;

/** Convert #RRGGBB or #RGB hex into an AWTRIX [R, G, B] tuple. */
export function hexToRgb(hex: string): [number, number, number] {
  const cleaned = hex.replace("#", "").trim();
  const full =
    cleaned.length === 3
      ? cleaned
          .split("")
          .map((c) => c + c)
          .join("")
      : cleaned;

  if (!/^[0-9a-fA-F]{6}$/.test(full)) {
    return [255, 107, 138];
  }

  const value = parseInt(full, 16);
  return [(value >> 16) & 255, (value >> 8) & 255, value & 255];
}

function normalizeBaseUrl(ipOrUrl: string): string {
  const trimmed = ipOrUrl.trim().replace(/\/+$/, "");
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  return `http://${trimmed}`;
}

async function postWithTimeout(
  url: string,
  body: unknown
): Promise<{ ok: boolean; status: number; error?: string; timedOut?: boolean }> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: controller.signal,
      cache: "no-store",
    });

    if (!response.ok) {
      const text = await response.text().catch(() => "");
      return {
        ok: false,
        status: response.status,
        error: text || `HTTP ${response.status}`,
      };
    }

    return { ok: true, status: response.status };
  } catch (error) {
    const isAbort =
      error instanceof Error &&
      (error.name === "AbortError" || error.message.includes("aborted"));

    return {
      ok: false,
      status: 0,
      timedOut: isAbort,
      error: isAbort
        ? `Device did not respond within ${REQUEST_TIMEOUT_MS}ms`
        : error instanceof Error
          ? error.message
          : "Unknown network error",
    };
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Try primary device URL first, then optional webhook / Tailscale relay.
 * Idempotent: same payload is sent to whichever endpoint succeeds first.
 */
export async function sendToAwtrix(
  config: Pick<DeviceConfig, "ip_or_url" | "webhook_relay_url">,
  path: string,
  payload: AwtrixCustomAppPayload | AwtrixNotifyPayload
): Promise<PushResult> {
  const endpoints: string[] = [];

  if (config.ip_or_url?.trim()) {
    endpoints.push(`${normalizeBaseUrl(config.ip_or_url)}${path}`);
  }

  if (config.webhook_relay_url?.trim()) {
    const relay = normalizeBaseUrl(config.webhook_relay_url);
    // Relay can be a full custom path or a base AWTRIX URL
    endpoints.push(
      relay.includes("/api/") ? relay : `${relay}${path}`
    );
  }

  if (endpoints.length === 0) {
    return {
      success: false,
      error: "No device IP/URL configured. Add one in Settings.",
    };
  }

  let lastError = "All endpoints failed";
  let timedOut = false;

  for (const endpoint of endpoints) {
    const result = await postWithTimeout(endpoint, payload);
    if (result.ok) {
      return { success: true, endpoint };
    }
    lastError = result.error || lastError;
    timedOut = Boolean(result.timedOut);
  }

  return { success: false, error: lastError, timedOut };
}

export function buildQuestionPayload(
  question: Pick<Question, "text" | "color_hex">,
  options?: {
    icon?: string;
    scrollSpeed?: number;
    duration?: number;
    repeat?: number;
  }
): AwtrixCustomAppPayload {
  const payload: AwtrixCustomAppPayload = {
    text: question.text.slice(0, 120),
    color: hexToRgb(question.color_hex || "#FF6B8A"),
    // omit icon unless provided — named icons like "heart" fail unless downloaded
    repeat: options?.repeat ?? -1,
    duration: options?.duration ?? 10,
    rainbow: false,
    wakeup: true,
  };

  if (options?.icon) payload.icon = options.icon;
  if (typeof options?.scrollSpeed === "number") {
    payload.scrollSpeed = options.scrollSpeed;
  }

  return payload;
}

export async function pushCustomApp(
  config: Pick<DeviceConfig, "ip_or_url" | "webhook_relay_url" | "scroll_speed">,
  question: Pick<Question, "text" | "color_hex">,
  appName = "daily_question"
): Promise<PushResult> {
  const payload = buildQuestionPayload(question, {
    scrollSpeed: config.scroll_speed,
  });

  const custom = await sendToAwtrix(
    config,
    `/api/custom?name=${encodeURIComponent(appName)}`,
    payload
  );
  if (custom.success) return custom;

  // Fallback: one-shot notification (always works even if custom apps glitch)
  const notify = await pushNotification(config, {
    text: payload.text,
    color: payload.color,
    duration: 12,
    wakeup: true,
    repeat: 3,
  });

  if (notify.success) {
    return {
      ...notify,
      error: `Custom app failed (${custom.error}); sent as notification instead`,
    };
  }

  return custom;
}

export async function pushNotification(
  config: Pick<DeviceConfig, "ip_or_url" | "webhook_relay_url">,
  payload: AwtrixNotifyPayload
): Promise<PushResult> {
  return sendToAwtrix(config, "/api/notify", payload);
}

export async function testConnection(
  config: Pick<DeviceConfig, "ip_or_url" | "webhook_relay_url">
): Promise<PushResult> {
  return pushNotification(config, {
    text: "Hello Love!",
    color: [255, 107, 138],
    duration: 8,
    wakeup: true,
    repeat: 2,
  });
}

File: awtrix.py
"""AWTRIX HTTP helpers for Daily Prompt."""

from __future__ import annotations

import re
from typing import Any
from urllib.parse import quote

import requests

TIMEOUT_S = 4.0


def hex_to_rgb(hex_color: str) -> list[int]:
    cleaned = hex_color.replace("#", "").strip()
    if len(cleaned) == 3:
        cleaned = "".join(c * 2 for c in cleaned)
    if not re.fullmatch(r"[0-9a-fA-F]{6}", cleaned):
        return [255, 107, 138]
    value = int(cleaned, 16)
    return [(value >> 16) & 255, (value >> 8) & 255, value & 255]


def normalize_base_url(ip_or_url: str) -> str:
    trimmed = ip_or_url.strip().rstrip("/")
    if re.match(r"^https?://", trimmed, re.I):
        return trimmed
    return f"http://{trimmed}"


def _post(url: str, payload: dict[str, Any]) -> tuple[bool, str]:
    try:
        response = requests.post(url, json=payload, timeout=TIMEOUT_S)
        if response.ok:
            return True, url
        body = (response.text or "").strip()
        return False, body or f"HTTP {response.status_code}"
    except requests.Timeout:
        return False, f"Device did not respond within {int(TIMEOUT_S)}s"
    except requests.RequestException as exc:
        return False, str(exc)


def send_to_awtrix(
    ip_or_url: str,
    path: str,
    payload: dict[str, Any],
    relay_url: str | None = None,
) -> tuple[bool, str]:
    endpoints: list[str] = []
    if ip_or_url.strip():
        endpoints.append(f"{normalize_base_url(ip_or_url)}{path}")
    if relay_url and relay_url.strip():
        relay = normalize_base_url(relay_url)
        endpoints.append(relay if "/api/" in relay else f"{relay}{path}")

    if not endpoints:
        return False, "No clock URL set. Add the IP in the sidebar."

    last_error = "All endpoints failed"
    for endpoint in endpoints:
        ok, detail = _post(endpoint, payload)
        if ok:
            return True, detail
        last_error = detail
    return False, last_error


def test_connection(ip_or_url: str, relay_url: str | None = None) -> tuple[bool, str]:
    return send_to_awtrix(
        ip_or_url,
        "/api/notify",
        {
            "text": "Hello Love!",
            "color": [255, 107, 138],
            "duration": 8,
            "wakeup": True,
            "repeat": 2,
        },
        relay_url=relay_url,
    )


def push_question(
    ip_or_url: str,
    text: str,
    color_hex: str = "#FF6B8A",
    relay_url: str | None = None,
    app_name: str = "daily_question",
) -> tuple[bool, str]:
    payload = {
        "text": text[:120],
        "color": hex_to_rgb(color_hex),
        "duration": 10,
        "repeat": -1,
        "rainbow": False,
        "wakeup": True,
    }
    ok, detail = send_to_awtrix(
        ip_or_url,
        f"/api/custom?name={quote(app_name)}",
        payload,
        relay_url=relay_url,
    )
    if ok:
        return True, detail

    # Fallback notification if custom app fails
    notify_ok, notify_detail = send_to_awtrix(
        ip_or_url,
        "/api/notify",
        {
            "text": text[:120],
            "color": hex_to_rgb(color_hex),
            "duration": 12,
            "wakeup": True,
            "repeat": 3,
        },
        relay_url=relay_url,
    )
    if notify_ok:
        return True, notify_detail
    return False, detail
